### PR TITLE
Add reachability check loop to expose reachability failures in node logs

### PR DESCRIPTION
**Pull request by pschork**
Default check interval is 5min
Minimum check interval is 10sec
Operators can disable by setting `NODE_REACHABILITY_POLL_INTERVAL` to 0

For backwards compatibility, the reachability check loop is disabled if the `NODE_DATAAPI_URL` is undefined (ie operator did not update their .env)

See also https://github.com/Layr-Labs/eigenda-operator-setup/pull/123

## Checks

- [x] I've made sure the lint is passing in this PR.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- Testing Strategy
   - [ ] Unit tests
   - [x] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by pschork** [comment]
Adds auto semantic versioning to node makefile.

```
> make semver
0.7.1-pschork-node-port-check.1
```

GitVersion knows that the latest tag release (currently) is `v0.7.0` and automatically bumps a patch release (ie `0.7.1`) and because this is a feature branch it includes branch name `pschork/node-port-check` plus build metadata `.1`

```
> make docker-node
> docker images opr-node
REPOSITORY   TAG                               IMAGE ID       CREATED          SIZE
opr-node     0.7.1-pschork-node-port-check.1   86d1ddd34fee   30 minutes ago   33.8MB
opr-node     latest                            86d1ddd34fee   30 minutes ago   33.8MB
```

```
> docker run -it opr-node:latest --version
da-node version 0.7.1-pschork-node-port-check.1-073ba2e-1715392025
```

File: node/node.go [hunk 1]
```diff
@@ -453,3 +458,76 @@ func (n *Node) checkCurrentNodeIp(ctx context.Context) {
...
+// OperatorReachabilityResponse is the response object for the reachability check
+type OperatorReachabilityResponse struct {
+	OperatorID      string `json:"operator_id"`
+	DispersalSocket string `json:"dispersal_socket"`
+	RetrievalSocket string `json:"retrieval_socket"`
+	DispersalOnline bool   `json:"dispersal_online"`
+	RetrievalOnline bool   `json:"retrieval_online"`
+}
+
+func (n *Node) checkNodeReachability() {
```
Comment (jianoaix) [inline node/node.go:471]: I'd build this into a metric (additionally log it as well)

File: node/node.go [hunk 2]
```diff
@@ -453,3 +458,76 @@ func (n *Node) checkCurrentNodeIp(ctx context.Context) {
...
+	}
+
+	n.Logger.Info("Start nodeReachabilityCheck goroutine in background to check the reachability of the operator node")
+	ticker := time.NewTicker(time.Duration(n.Config.ReachabilityPollIntervalSec) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		<-ticker.C
+
+		if n.Config.DataApiUrl == "" {
```
Comment (ian-shim) [inline node/node.go:484]: can we early exit this method instead of spinning up a new goroutine if dataapi url is empty?

File: node/node.go [hunk 3]
```diff
@@ -453,3 +458,76 @@ func (n *Node) checkCurrentNodeIp(ctx context.Context) {
...
+			n.Logger.Error("Unable to perform reachability check - NODE_DATAAPI_URL is not defined in .env")
+			continue
+		}
+
+		checkUrl, err := url.Parse(fmt.Sprintf("%s/api/v1/operators-info/port-check?operator_id=%s", n.Config.DataApiUrl, n.Config.ID.Hex()))
+		if err != nil {
+			n.Logger.Error("Reachability check failed - invalid check url", err, "checkUrl", checkUrl.String())
```
Comment (ian-shim) [inline node/node.go:491]: Do we want to return in this case or log error and continue checking reachability?

File: node/node.go [hunk 4]
```diff
@@ -453,3 +458,76 @@ func (n *Node) checkCurrentNodeIp(ctx context.Context) {
...
+	for {
+		<-ticker.C
+
+		if n.Config.DataApiUrl == "" {
+			n.Logger.Error("Unable to perform reachability check - NODE_DATAAPI_URL is not defined in .env")
+			continue
+		}
+
+		checkUrl, err := url.Parse(fmt.Sprintf("%s/api/v1/operators-info/port-check?operator_id=%s", n.Config.DataApiUrl, n.Config.ID.Hex()))
+		if err != nil {
```
Comment (ian-shim) [inline node/node.go:490]: Which condition does it fall into if the dataapi itself was down?

File: node/node.go (hunk 2 above)
Comment (pschork) [reply node/node.go:484]: My original implementation did that, but I changed the implementation to continuously throw errors to increase the likelihood of operators seeing the error. This error condition will only occur if the operator updates their docker image without manually adding the `NODE_DATAAPI_URL`.

File: node/node.go (hunk 3 above)
Comment (pschork) [reply node/node.go:491]: My original implementation did that, but I changed the implementation to continuously throw errors to increase the likelihood of operators seeing the error.

File: node/node.go (hunk 4 above)
Comment (pschork) [reply node/node.go:490]: Not this one. This is just ensuring that the constructed URL parses as valid.
If DataAPI is [down/unresolvable
](https://github.com/Layr-Labs/eigenda/pull/556/files#diff-375d57e386f20eaa5f09f02bb9d28bfc48ac3dca18d0325f59492208219e5618R498)
If DataAPI is [up but throwing non-200  ](https://github.com/Layr-Labs/eigenda/pull/556/files#diff-375d57e386f20eaa5f09f02bb9d28bfc48ac3dca18d0325f59492208219e5618R504)

File: node/node.go (hunk 3 above)
Comment (ian-shim) [reply node/node.go:491]: I think this exits instead of continuously throwing error

Comment (pschork) [reply node/node.go]: Good catch.

Comment (pschork) [reply node/node.go]: Reverted to original disabled logic

File: node/node.go (hunk 2 above)
Comment (pschork) [reply node/node.go]: We early exit now

File: node/flags/flags.go [hunk 5]
```diff
@@ -181,6 +181,21 @@ var (
 		Value:    "180",
 		EnvVar:   common.PrefixEnvVar(EnvVarPrefix, "EXPIRATION_POLL_INTERVAL"),
 	}
+	ReachabilityPollIntervalSecFlag = cli.StringFlag{
+		Name:     common.PrefixFlag(FlagPrefix, "reachability-poll-interval"),
+		Usage:    "How often (in second) to check if node is reachabile from DA backend",
```
Comment (jianoaix) [inline node/flags/flags.go:186]: nit: "DA backend" -> "Disperser"

File: node/flags/flags.go [hunk 6]
```diff
@@ -181,6 +181,21 @@ var (
...
+		Required: false,
+		Value:    "300",
```
Comment (jianoaix) [inline node/flags/flags.go:188]: should we do it a bit more frequent, like once per min? It needs to clearly more frequent than batch interval to be more useful

File: node/node.go [hunk 7]
```diff
@@ -453,3 +458,80 @@ func (n *Node) checkCurrentNodeIp(ctx context.Context) {
...
+	ticker := time.NewTicker(time.Duration(n.Config.ReachabilityPollIntervalSec) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		<-ticker.C
+
+		n.Logger.Debug("Calling reachability check", "url", checkUrl.String())
+
+		resp, err := http.Get(checkUrl.String())
+		if err != nil {
```
Comment (jianoaix) [inline node/node.go:498]: Should we set a value in the gauge in case it gets error? @shrimalmadhur

**Review by jianoaix — APPROVED** [review]
LGTM, just one comment on if we need to update the gauge, when the call to dataapi fails to get response. @shrimalmadhur may have dealt with an issue for the onchain metric before, please chime in.

**Review by ian-shim — APPROVED** [review]

**Comment by pschork** [comment]
> LGTM, just one comment on if we need to update the gauge, when the call to dataapi fails to get response. @shrimalmadhur may have dealt with an issue for the onchain metric before, please chime in.

Thanks @jianoaix. If Dataapi is down, this isn't really the operators issue and does not mean they are unreachable. We will log this condition as error in log every interval, so I feel like it wont go unnoticed.

File: node/node.go (hunk 7 above)
Comment (shrimalmadhur) [reply node/node.go:498]: maybe we should, otherwise it will show last value.

**Comment by shrimalmadhur** [comment]
> > LGTM, just one comment on if we need to update the gauge, when the call to dataapi fails to get response. @shrimalmadhur may have dealt with an issue for the onchain metric before, please chime in.
> 
> Thanks @jianoaix. If Dataapi is down, this isn't really the operators issue and does not mean they are unreachable. We will log this condition as error in log every interval, so I feel like it wont go unnoticed.

so if the dataapi is down - what's the indicator for operators? like they will still see reachable from their side since the gauge value has not changed. How do they know if data api is down and reachability check is not working fine?

**Pull request merged**